tests/js/server/aql/aql-index-cache.js

```diff
@@ -0,0 +1,996 @@
+        db._drop(cn);
+      }
+    }
+  };
+}
+
+function FiguresSuite () {
+  let docs = [];
+  for (let i = 0; i < 1000; ++i) {
+    docs.push({ value: i });
+  }
+
+  let checkIndex = (options, cacheEnabled) => {
+    options.name = "testCache";
+    if (cacheEnabled !== undefined) {
+      options.cacheEnabled = cacheEnabled;
+    }
+
+    let c = db._create(cn);
+    try {
+      let idx = c.ensureIndex(options);
+      assertEqual(options.type, idx.type);
+      assertTrue(idx.hasOwnProperty("cacheEnabled"));
+      assertEqual(Boolean(cacheEnabled), idx.cacheEnabled);
+
+      if (!isCluster) {
+        // cluster unfortunately does not report any index figures
+        // (and never did)
+        let figures = c.indexes(true).filter((idx) => idx.name === 'testCache')[0];
+        if (!cacheEnabled) {
+          assertFalse(figures.figures.cacheInUse, figures);
+          assertEqual(0, figures.figures.cacheSize, figures);
+          assertEqual(0, figures.figures.cacheUsage, figures);
+          assertFalse(figures.figures.hasOwnProperty("cacheLifeTimeHitRate"), figures);
+          assertFalse(figures.figures.hasOwnProperty("cacheWindowedTimeHitRate"), figures);
+          } else {
+          assertTrue(figures.figures.cacheInUse, figures);
+          assertNotEqual(0, figures.figures.cacheSize, figures);
+          assertEqual(0, figures.figures.cacheUsage, figures);
+          assertTrue(figures.figures.cacheSize > figures.figures.cacheUsage, figures);
+          assertTrue(figures.figures.hasOwnProperty("cacheLifeTimeHitRate"), figures);
+          assertTrue(figures.figures.hasOwnProperty("cacheWindowedHitRate"), figures);
+          assertEqual(0, figures.figures.cacheLifeTimeHitRate, figures);
+          assertEqual(0, figures.figures.cacheWindowedHitRate, figures);
+        }
+
+        if (cacheEnabled) {
+          c.insert(docs);
+      
+          // figures shouldn't have changed
+          let figures = c.indexes(true).filter((idx) => idx.name === 'testCache')[0];
+          assertTrue(figures.figures.cacheInUse, figures);
+          assertNotEqual(0, figures.figures.cacheSize, figures);
+          assertEqual(0, figures.figures.cacheUsage, figures);
+          assertTrue(figures.figures.hasOwnProperty("cacheLifeTimeHitRate"), figures);
+          assertTrue(figures.figures.hasOwnProperty("cacheWindowedHitRate"), figures);
+          assertEqual(0, figures.figures.cacheLifeTimeHitRate, figures);
+          assertEqual(0, figures.figures.cacheWindowedHitRate, figures);
+
+          // figures should now change due to reading the data into the cache
+          for (let i = 0; i < 5; ++i) {
+            db._query(`FOR i IN 1..1000 FOR doc IN ${cn} FILTER doc.value == i RETURN 1`);
+          }
+          
+          figures = c.indexes(true).filter((idx) => idx.name === 'testCache')[0];
+          assertNotEqual(0, figures.figures.cacheSize, figures);
+          assertNotEqual(0, figures.figures.cacheUsage, figures);
```

Comment:
Unfortunately this cannot be guaranteed.
It is always possible that a value couldn't be inserted into the cache, e.g. if cache can be locked because it is currently being migrated, or an insert can be rejected if the combined memory usage of all caches exceeds the system-wide limit etc.
In most cases asserting that the cache usage or hit rate is increasing will work here, but not in some edge cases (Jenkins under memory pressure, cache migration, ...), so we could see some false positives during testing here if we add the asserts.